### PR TITLE
docs: add discovery protocol doc

### PR DESCRIPTION
**Pull request by yichengq**
This document talks about the technical details of discovery service
protocol. It helps users to learn about how discovery service works and
what behavior to expect.

fixes https://github.com/coreos/etcd/issues/2095

### CODE REVIEW

File: Documentation/discovery_protocol.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+This document talks about the technical details of discovery service protocol. It helps users to learn about how discovery service works and what behavior to expect.
+
+# Discovery Service Protocol
+
+Discovery service protocol helps new etcd member to discover all other members in cluster bootstrap phase using a shared discovery URL.
+
+Discovery service protocol is _only_ used in cluster bootstrap phase, and cannot be used for runtime reconfiguration or cluster monitoring.
+
```
Comment (xiang90) [inline Documentation/discovery_protocol.md]: we also need to mention you need a new token if the previous token is used (even if the bootstrap failed.)

Comment (yichengq) [inline Documentation/discovery_protocol.md]: added it below

File: Documentation/discovery_protocol.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
+This document talks about the technical details of discovery service protocol. It helps users to learn about how discovery service works and what behavior to expect.
```
Comment (xiang90) [inline Documentation/discovery_protocol.md]: we need to give this doc a title at the first line

Comment (yichengq) [inline Documentation/discovery_protocol.md]: i will put the total title at the first line

**Comment by xiang90** [comment]
LGTM

/cc @robszumski Can you please do a double checking?

File: Documentation/discovery_protocol.md (hunk 2 above)
Comment (yichengq) [inline Documentation/discovery_protocol.md]: It says the same thing as line `The rest of this article will walk...`. I will delete this one.

File: Documentation/discovery_protocol.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+CoreOS Inc. hosts a public discovery service at https://discovery.etcd.io/ , which provides some nice features for ease of use.
+
+### Mask Key Prefix
+
+Public discovery service will redirect `https://discovery.etcd.io/${UUID}` to etcd cluster behind for the key at `/v2/keys/_etcd/registry`. It masks register key prefix for short and readable discovery url.
+
+### Get new token
+
+GET /new
```
Comment (robszumski) [inline Documentation/discovery_protocol.md]: `GET /new`

File: Documentation/discovery_protocol.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	400 Bad Request
+200 Body:
+	generated discovery url
+```
+
+The generation process in the service follows the step from [Creating a New Discovery Token](#creating-a-new-discovery-token) to [Specifying the Expected Cluster Size](#specifying-the-expected-cluster-size).
+
+### Check Discovery Status
+
+GET /${UUID}
```
Comment (robszumski) [inline Documentation/discovery_protocol.md]:
```
GET /${UUID}
```

File: Documentation/discovery_protocol.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+You could check discovery status for this token so far, including the machines that have been registered.
```
Comment (robszumski) [inline Documentation/discovery_protocol.md]: "You can check the status for this discovery token, including the machines that have been registered, by requesting the value of the UUID."

**Comment by yichengq** [comment]
@robszumski Have addressed all your comments. PTAL

**Comment by robszumski** [comment]
LGTM

**Pull request merged**